### PR TITLE
Adding frame/iframe switching support

### LINKED ISSUE #120: Enhancement Request: Add steps to allow Webdriver to switch to frames/iframes
Now, Iridium supports switching to new tab, new window and new alert. In some cases, have to handle the elements within iframes.

### PR DESCRIPTION
**Pull request by mdjebaile**
This PR resolves #120 and adds step functionality for frames/iframes

### CODE REVIEW

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (nesvand) [inline src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: contains the

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (nesvand) [inline src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: no docs?

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (nesvand) [inline src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: `'`typo?

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (mdjebaile) [reply src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: actually, do we need this functionality?

**Review by nesvand — COMMENTED** [review]
Is error-handling performed at a different level? ie. if a selector doesn't find something? I haven't dived deep enough to know where we should be handling exceptions

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (mdjebaile) [reply src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: c&p error, fixed

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (mdjebaile) [reply src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: c&p error, fixed

File: src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java
Comment (nesvand) [reply src/main/java/au/com/agic/apptesting/steps/FrameStepDefinition.java]: how are the frame sets created in the first place? if it's not possible at this time then we don't need to switch between them - this feature can be added then (ie. remove this step definition if we can't make frame sets)

**Review by nesvand — CHANGES_REQUESTED** [review]
Waiting on feedback and any necessary changes before merging

**Comment by mdjebaile** [comment]
@nesvand exception handling is very trivial ATM, catching all exceptions the same way (without recovering)

**Comment by nesvand** [comment]
@mdjebaile that's a shame - maybe worth keeping track of as an enhancement, but for now if this is expected behaviour I'll continue on to local testing ✔️

**Comment by nesvand** [comment]
Also @mdjebaile - I've noticed the zap plugin is deleting itself for you too - is this going to be problematic? cc: @mcasperson

**Comment by mcasperson** [comment]
The deleted ZAP plugin could be an issue. Was it removed for a reason?

**Comment by nesvand** [comment]
image

[screen-recording.zip](https://github.com/AutoGeneral/IridiumApplicationTesting/files/1757756/screen-recording.zip)

```
Feature: Iridium - can focus CORS and non-CORS iframes, interact and collect data

    Scenario: Generate Page Object
        Given the alias mappings
            | CORS_id | cors |
            | CORS_class | iframe__cors |
            | CORS_name | cors-iframe |
            | CORS_selector | .iframe.iframe__cors |
            | CORS_xpath | //*[@id="cors"] |
            | non_CORS_id | non-cors |
            | non_CORS_class | iframe__non-cors |
            | non_CORS_name | non-cors-iframe |
            | non_CORS_selector | .iframe.iframe__non-cors |
            | non_CORS_xpath | //*[@id="non-cors"] |
            | submit_button | submit-button |
            | click_content | click-content |
            | CORS_content | cors submit clicked |
            | non_CORS_content | non-cors submit clicked |
            | lorem_ipsum | Lorem ipsum dolor sit amet |

    Scenario: Default Profile
        When I set the default wait time between steps to "0.5" seconds
        And I set the default wait for elements to be available to "1" second
        And I set the default keystroke delay to "0" milliseconds
        And I delete all cookies

    Scenario: Prepare the data
        And I dump the alias map to the console

    Scenario: Start iframe test application
        And I open the application
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed

    Scenario: Select CORS iframe with id
        When I switch to the iframe with the ID alias of "CORS_id"
        And I verify that the page does not contain the text alias "CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "CORS_content"
        And I switch to the default content

    Scenario: Select non-CORS iframe with id
        When I switch to the iframe with the ID alias of "non_CORS_id"
        And I verify that the page does not contain the text alias "non_CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "non_CORS_content"
        And I switch to the default content
        And I verify that the page contains the text alias "lorem_ipsum"
        And I refresh the page
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed

    Scenario: Select CORS iframe with class
        When I switch to the iframe with the class alias of "CORS_class"
        And I verify that the page does not contain the text alias "CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "CORS_content"
        And I switch to the default content

    Scenario: Select non-CORS iframe with class
        When I switch to the iframe with the class alias of "non_CORS_class"
        And I verify that the page does not contain the text alias "non_CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "non_CORS_content"
        And I switch to the default content
        And I verify that the page contains the text alias "lorem_ipsum"
        And I refresh the page
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed

    Scenario: Select CORS iframe with name
        When I switch to the iframe with the name alias of "CORS_name"
        And I verify that the page does not contain the text alias "CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "CORS_content"
        And I switch to the default content

    Scenario: Select non-CORS iframe with name
        When I switch to the iframe with the name alias of "non_CORS_name"
        And I verify that the page does not contain the text alias "non_CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "non_CORS_content"
        And I switch to the default content
        And I verify that the page contains the text alias "lorem_ipsum"
        And I refresh the page
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed

    Scenario: Select CORS iframe with selector
        When I switch to the iframe with the css selector alias of "CORS_selector"
        And I verify that the page does not contain the text alias "CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "CORS_content"
        And I switch to the default content

    Scenario: Select non-CORS iframe with selector
        When I switch to the iframe with the css selector alias of "non_CORS_selector"
        And I verify that the page does not contain the text alias "non_CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "non_CORS_content"
        And I switch to the default content
        And I verify that the page contains the text alias "lorem_ipsum"
        And I refresh the page
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed

    Scenario: Select CORS iframe with xpath
        When I switch to the iframe with the xpath alias of "CORS_xpath"
        And I verify that the page does not contain the text alias "CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "CORS_content"
        And I switch to the default content

    Scenario: Select non-CORS iframe with xpath
        When I switch to the iframe with the xpath alias of "non_CORS_xpath"
        And I verify that the page does not contain the text alias "non_CORS_content"
        And I click the element found by alias "submit_button"
        And I verify that the page contains the text alias "non_CORS_content"
        And I switch to the default content
        And I verify that the page contains the text alias "lorem_ipsum"
        And I refresh the page
        And I wait "3" seconds for the element found by alias "CORS_id" to be displayed
```

**Comment by nesvand** [comment]
@mcasperson For whatever reason IntelliJ insists on deleting that particular plugin at every opportunity it gets

I'll restore the file for now

**Review by nesvand — APPROVED** [review]

**Comment by nesvand** [comment]
@mcasperson looks like it's being deleted by Symantic Endpoint Protection - silly virus scanners 💢 

@mdjebaile add an exception for the file - no idea why the virus scanner is so angry at the fuzzer plugin and not the others 🤷‍♂️

**Comment by srowatt** [comment]
The fuzzdb plugin will keep getting deleted because it contains a list of attach patterns that will be detected by virus scanners as a virus. As Andrew said you'll need to add it to the list of exclusions for virus scanning.

**Review by maraisr — APPROVED** [review]

**Comment by mdjebaile** [comment]
@nesvand it seems I don't have permissions to merge, you can go ahead and do it for me

**Pull request merged**
